File: src/asset.py
from pathlib import Path

import yaml

from src.config import ROOT


class AssetStore:
    def __init__(self, run_id: str):
        self.run_dir = ROOT / "runs" / run_id
        self.run_dir.mkdir(parents=True, exist_ok=True)

    def save(self, stage: str, name: str, data: dict | list) -> Path:
        stage_dir = self.run_dir / stage
        stage_dir.mkdir(exist_ok=True)
        path = stage_dir / f"{name}.yaml"
        with open(path, "w") as f:
            yaml.dump(data, f, allow_unicode=True, default_flow_style=False)
        return path

    def load(self, stage: str, name: str) -> dict | list:
        path = self.run_dir / stage / f"{name}.yaml"
        with open(path) as f:
            return yaml.safe_load(f)

    def save_binary(self, stage: str, name: str, data: bytes) -> Path:
        stage_dir = self.run_dir / stage
        stage_dir.mkdir(exist_ok=True)
        path = stage_dir / name
        with open(path, "wb") as f:
            f.write(data)
        return path

    def log_input(self, stage: str, data: dict) -> None:
        self.save(stage, "input", data)

    def log_output(self, stage: str, data: dict) -> None:
        self.save(stage, "output", data)

    def audio_dir(self) -> Path:
        d = self.run_dir / "audio"
        d.mkdir(exist_ok=True)
        return d

    def video_dir(self) -> Path:
        d = self.run_dir / "video"
        d.mkdir(exist_ok=True)
        return d

File: src/asset.ts

import path from "path";
import fs from "fs-extra";
import yaml from "js-yaml";
import { ROOT } from "./config.js";

export class AssetStore {
    runDir: string;

    constructor(runId: string) {
        this.runDir = path.join(ROOT, "runs", runId);
        fs.ensureDirSync(this.runDir);
    }

    loadState(): any {
        const p = path.join(this.runDir, "state.json");
        return fs.existsSync(p) ? JSON.parse(fs.readFileSync(p, "utf-8")) : {};
    }

    updateState(update: any): void {
        const state = this.loadState();
        const newState = { ...state, ...update };
        fs.writeFileSync(path.join(this.runDir, "state.json"), JSON.stringify(newState, null, 2));
    }

    save(stage: string, name: string, data: any): string {
        const stageDir = path.join(this.runDir, stage);
        fs.ensureDirSync(stageDir);
        const p = path.join(stageDir, `${name}.yaml`);
        fs.writeFileSync(p, yaml.dump(data));
        return p;
    }

    load(stage: string, name: string): any {
        const p = path.join(this.runDir, stage, `${name}.yaml`);
        return yaml.load(fs.readFileSync(p, "utf8"));
    }

    saveBinary(stage: string, name: string, data: Buffer): string {
        const stageDir = path.join(this.runDir, stage);
        fs.ensureDirSync(stageDir);
        const p = path.join(stageDir, name);
        fs.writeFileSync(p, data);
        return p;
    }

    logInput(stage: string, data: any): void {
        this.save(stage, "input", data);
    }

    logOutput(stage: string, data: any): void {
        this.save(stage, "output", data);
    }

    audioDir(): string {
        const d = path.join(this.runDir, "audio");
        fs.ensureDirSync(d);
        return d;
    }

    videoDir(): string {
        const d = path.join(this.runDir, "video");
        fs.ensureDirSync(d);
        return d;
    }
}
